### PR TITLE
test(vpc,transit_gateway): add native OpenTofu test coverage

### LINKED ISSUE #384: bug(vpc): NAT-gateway routes and public route-table association crash plan when enable_internet_gateway=false or public_subnets_list is empty
## Affected module path
`modules/aws/vpc`

## OpenTofu / provider versions
- OpenTofu v1.12.4
- hashicorp/aws provider v6.54.0 (module requires `>= 6.0.0`)

## Summary
Several resources in `modules/aws/vpc/main.tf` compute their `count` from `var.enable_nat_gateway` and/or subnet-list length alone, without also checking `local.enable_igw` (which is `var.enable_internet_gateway && length(var.public_subnets_list) != 0`). Since NAT gateways themselves (`aws_nat_gateway.natgw`) are correctly gated on `local.enable_igw`, this mismatch causes a plan-time crash whenever a caller sets `enable_internet_gateway = false` (or supplies an empty `public_subnets_list`) while leaving `enable_nat_gateway` at its default `true`.

Affected resources (their `count` expressions do not account for `local.enable_igw`):
- `aws_route.private_default_route_natgw` (main.tf ~line 312-317)
- `aws_route.db_default_route_natgw` (main.tf ~line 333-338)
- `aws_route.dmz_default_route_natgw` (main.tf ~line 354-359)
- `aws_route.mgmt_default_route_natgw` (main.tf ~line 375-380)
- `aws_route.workspaces_default_route_natgw` (main.tf ~line 396-401)
- `aws_route_table_association.public` (main.tf ~line 418-422), which indexes `aws_route_table.public_route_table[0]` unconditionally even though that resource's own `count` is gated on `local.enable_igw`

## Repro steps
From `modules/aws/vpc`, run `tofu test` against a plan with:
```hcl
name                    = "core-vpc"
enable_internet_gateway = false
```
(all other variables left at their defaults, so `enable_nat_gateway = true` and `private_subnets_list`/`public_subnets_list` are non-empty).

A minimal reproducing test case is included (commented, not asserting a passing plan) in the PR that reported this issue: see `modules/aws/vpc/tests/vpc.tftest.hcl`, run `enable_internet_gateway_false_disables_igw_and_dependent_nat`.

## Expected vs actual behavior
- Expected: with `enable_internet_gateway = false`, the module should still produce a valid plan. Since NAT gateways require an IGW/public subnet to attach to, the NAT-gateway-dependent routes (and the public route-table association) should also be skipped (count 0) in this configuration, consistent with `aws_nat_gateway.natgw` itself being skipped.
- Actual: `tofu plan`/`tofu test` crashes with `Call to function "element" failed: cannot use element function with an empty list.` (for the `*_default_route_natgw` resources referencing `aws_nat_gateway.natgw[*].id`) and `The given key does not identify an element in this collection value: the collection has no elements.` (for `aws_route_table_association.public` referencing `aws_route_table.public_route_table[0]`).

## Actual tofu test error output (excerpt)
```
Error in function call

  on main.tf line 315, in resource "aws_route" "private_default_route_natgw":
 315:   nat_gateway_id         = element(aws_nat_gateway.natgw[*].id, count.index)
    +-------------------
    | while calling element(list, index)
    | aws_nat_gateway.natgw is empty tuple
    | count.index is 1

Call to function "element" failed: cannot use element function with an empty list.
```
```
Error: Invalid index

  on main.tf line 420, in resource "aws_route_table_association" "public":
 420:   route_table_id = aws_route_table.public_route_table[0].id

The given key does not identify an element in this collection value: the collection has no elements.
```

## Suggested acceptance criteria
- The `count` expressions for the 5 `*_default_route_natgw` resources incorporate the same "NAT gateway actually created" condition as `aws_nat_gateway.natgw` itself (i.e. also require `local.enable_igw`), so they are skipped when NAT gateways are skipped.
- `aws_route_table_association.public`'s `count` also incorporates `local.enable_igw`, so it is skipped when the public route table itself is skipped.
- A `tofu test` case with `enable_internet_gateway = false` (NAT gateway left at its default-enabled value) plans successfully with 0 NAT-gateway-dependent routes and 0 public route-table associations.
- A `tofu test` case with `public_subnets_list = []` (NAT gateway left enabled) plans successfully for the same reason.

## Reported by
Automated native-test-coverage effort (see PR adding `modules/aws/vpc/tests/`).

### PR DESCRIPTION
**Pull request by zachreborn**
# Description
Adds native OpenTofu test coverage (`tests/*.tftest.hcl`) for `modules/aws/vpc` and all four `modules/aws/transit_gateway/*` submodules (`attachment`, `connect`, `route`, `tgw`), per `AGENTS.md` § Module Design Specifications > 6. Native Test Coverage. This is one of six parallel PRs splitting the full 28-module test-coverage effort by module group; see the run link below for the overall plan.

This PR is test-only — it does not modify any `main.tf`/`outputs.tf`/`variables.tf`. Writing full toggle coverage surfaced a few real module bugs; those are split out into a separate PR: **#409 (`fix(vpc): gate NAT-gateway resources on IGW state`, fixes #384)**, which also fixes a typo in `transit_gateway/connect`'s `configurations` output and adds wiring-testable outputs to `flow_logs`. Until #409 merges, the handful of tests here that touch those scenarios use documented workarounds/current-key-names (see inline comments referencing #384/#409) so this suite stays green against the module code as it exists on `main` today.

Coverage added:
- **vpc**: baseline plan + full value-level output assertions; NAT gateway toggles (`enable_nat_gateway`, `single_nat_gateway`); Internet Gateway toggle including the empty-`public_subnets_list` edge case; SSM/ECR/S3 VPC endpoint toggles; firewall route toggles; an IPAM-sourced CIDR run; `vpc_flow_logs` module wiring; the CloudWatch Internet Monitor toggle, its `monitor_name` precondition, and its dynamic S3 delivery block (with value-level assertions); and one `run` block per `validation{}` block in `variables.tf` (10 validation blocks, several with both an above-range/below-range or upper/lower-bound failure case).
- **transit_gateway/attachment**: `for_each` over `vpc_ids` (including the empty-map case), per-entry defaults/overrides, value-level `info`/`vpc_owner_id` output assertions, and `enable_flow_logs` module wiring.
- **transit_gateway/connect**: connect attachment + connect peer `for_each`, defaults/overrides, value-level output assertions, an `override_resource`-based sentinel proving `transit_gateway_address` is left for the provider rather than coerced, and the `protocol` validation block.
- **transit_gateway/route**: `for_each` over `destination_cidr_blocks`, the `blackhole` toggle, and the empty-set edge case.
- **transit_gateway/tgw**: baseline defaults/overrides, tag merging, full value-level output assertions, and `enable_flow_logs` module wiring.

All tests run fully offline via `mock_provider`/`mock_resource`/`override_resource` (`tofu init -backend=false && tofu test`) — no AWS credentials or backend required. Several test files add `mock_resource` ARN-format defaults for `aws_iam_role`/`aws_iam_policy`/`aws_cloudwatch_log_group`/`aws_vpc`, since the nested `flow_logs` module and the Internet Monitor resource wire computed ARNs into arguments the AWS provider schema validates as well-formed ARNs, and mock_provider's generic placeholder strings aren't ARN-shaped.

## Issue or Ticket
Refs #384 (fix split out to #409)

## Type of change
- [x] `test` — adding or correcting tests

## Breaking Changes
- [ ] Yes — this is a breaking change
- [x] No

## TODOs
- [x] PR title follows Conventional Commits (`<type>(scope): description`).
- [x] Validate your code matches the style of the project (`tofu fmt -recursive`).
- [ ] Update the docs (regenerate the `terraform-docs` block where applicable) — not applicable, no README changes in this PR.
- [x] Validate all tests run successfully, including pre-commit checks.
- [x] Include release notes and description.

## Testing
Ran `tofu init -backend=false && tofu test` in each of the 5 module directories (against the module code as it exists on `main` today, i.e. without #409):
- `modules/aws/vpc`: 38 passed, 0 failed
- `modules/aws/transit_gateway/attachment`: 6 passed, 0 failed
- `modules/aws/transit_gateway/connect`: 8 passed, 0 failed
- `modules/aws/transit_gateway/route`: 4 passed, 0 failed
- `modules/aws/transit_gateway/tgw`: 5 passed, 0 failed

`tofu fmt -check -diff -recursive` is clean across all changed files.

---
Conversation: https://app.warp.dev/conversation/b2ae68bc-4421-4302-9406-2aa93e72b298
Run: https://oz.warp.dev/runs/019f5e3e-7d57-710f-a723-6b0ddf7d4d72

Co-Authored-By: Oz <oz-agent@warp.dev>

### CODE REVIEW

File: modules/aws/vpc/tests/vpc.tftest.hcl [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+# a real module bug, filed as https://github.com/zachreborn/terraform-modules/issues/384
+# -- not a test bug, so it is not being worked around by editing main.tf here.
+run "enable_internet_gateway_false_disables_igw_and_dependent_nat" {
+  command = plan
+
+  variables {
+    name                    = "core-vpc"
+    enable_internet_gateway = false
+    enable_nat_gateway      = false
+    public_subnets_list     = []
```
[the remark this answers is not in the document]
Comment (zachreborn) [reply modules/aws/vpc/tests/vpc.tftest.hcl:297]: Fixed in 669c6294: added local.enable_natgw (var.enable_nat_gateway && local.enable_igw) and gated aws_eip.nateip/aws_nat_gateway.natgw/the five *_default_route_natgw routes on it, and gated aws_route_table_association.public on local.enable_igw. The regression test in 6e3acf36 now runs with the default enable_nat_gateway=true and non-empty public_subnets_list, and asserts zero NAT/route resources instead of masking the scenario.

File: modules/aws/vpc/tests/vpc.tftest.hcl [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+
+# See the note above run "enable_internet_gateway_false_disables_igw_and_dependent_nat":
+# enable_nat_gateway is disabled here for the same reason (tracked in issue #384).
+run "empty_public_subnets_list_disables_igw_even_when_enabled" {
+  command = plan
+
+  variables {
+    name                    = "core-vpc"
+    enable_internet_gateway = true
+    enable_nat_gateway      = false
```
[the remark this answers is not in the document]
Comment (zachreborn) [reply modules/aws/vpc/tests/vpc.tftest.hcl:359]: Fixed in 669c6294 (same root-cause fix as the sibling comment above): local.enable_natgw now gates NAT gateway/EIP/route creation on local.enable_igw. The empty-public-subnets-list run in 6e3acf36 now leaves enable_nat_gateway at its default true and asserts the NAT gateway/EIP/private route are correctly skipped rather than avoiding the scenario.

File: modules/aws/vpc/tests/validation.tftest.hcl [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+  command = plan
+
+  variables {
+    name             = "core-vpc"
+    enable_flow_logs = false
+    subnet_indices   = [3]
+  }
+
+  expect_failures = [var.subnet_indices]
+}
```
[the remark this answers is not in the document]
Comment (zachreborn) [reply modules/aws/vpc/tests/validation.tftest.hcl]: Fixed in 6e3acf36: added rejects_subnet_index_below_range (subnet_indices = [-1]) alongside the existing above-range case.

File: modules/aws/vpc/tests/internet_monitor.tftest.hcl [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    internet_monitor_monitor_name     = "core-vpc-monitor"
+    internet_monitor_s3_bucket_name   = "core-vpc-internet-monitor-logs"
+    internet_monitor_s3_bucket_prefix = "internet-monitor/"
+    internet_monitor_s3_bucket_status = "ENABLED"
+  }
+
+  assert {
+    condition     = length(aws_internetmonitor_monitor.this[0].internet_measurements_log_delivery) == 1
+    error_message = "The dynamic internet_measurements_log_delivery block should be populated when internet_monitor_s3_bucket_name is set."
+  }
```
[the remark this answers is not in the document]
Comment (zachreborn) [reply modules/aws/vpc/tests/internet_monitor.tftest.hcl:109]: Fixed in 6e3acf36: now asserts the s3_config bucket_name, bucket_prefix, and log_delivery_status values directly, not just that the block exists.

File: modules/aws/vpc/tests/vpc.tftest.hcl [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+run "baseline_plans_with_defaults" {
+  command = plan
+
+  variables {
+    name = "core-vpc"
+  }
+
+  assert {
+    condition     = aws_vpc.vpc.cidr_block == var.vpc_cidr
+    error_message = "The VPC's cidr_block should equal the (default) vpc_cidr input."
```
[the remark this answers is not in the document]
Comment (zachreborn) [reply modules/aws/vpc/tests/vpc.tftest.hcl:35]: Fixed in 6e3acf36: added run "ipam_sourced_cidr_passes_ipam_arguments_through", which sets ipv4_ipam_pool_id/ipv4_netmask_length and asserts both pass through unchanged. Note: cidr_block itself is Optional+Computed on aws_vpc, so once ipv4_ipam_pool_id is set, both real Terraform and the mock provider resolve a concrete value for it rather than leaving it null (mirroring real apply-time behavior where it would show as "(known after apply)") -- so that specific attribute isn't asserted as null, but the two IPAM arguments are directly asserted, which is the meaningful, regression-catching proof that the module took the IPAM branch.

File: modules/aws/vpc/tests/vpc.tftest.hcl [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+  }
+
+  assert {
+    condition     = length(aws_internetmonitor_monitor.this) == 0
+    error_message = "Internet Monitor should not be created by default (enable_internet_monitor=false)."
+  }
+
+  assert {
+    condition     = output.vpc_id != null
+    error_message = "vpc_id output should resolve."
```
[the remark this answers is not in the document]
Comment (zachreborn) [reply modules/aws/vpc/tests/vpc.tftest.hcl]: Fixed in 6e3acf36: the baseline run now compares every output (vpc_id, vpc_cidr_block, vpc_arn, default_security_group_id, all six subnet_ids outputs, private/public subnet CIDRs and ARNs, availability_zone, all six route table id outputs, nat_eips, nat_eips_public_ips, natgw_ids, igw_id, name) directly against the corresponding resource attribute instead of just checking non-nullness/cardinality.

File: modules/aws/transit_gateway/connect/tests/connect.tftest.hcl [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+  }
+
+  assert {
+    condition     = length(output.arns) == 2
+    error_message = "arns output should contain one entry per connect peer."
+  }
+
+  assert {
+    condition     = length(output.configurations) == 2
+    error_message = "configurations output should contain one entry per connect peer."
```
[the remark this answers is not in the document]
Comment (zachreborn) [reply modules/aws/transit_gateway/connect/tests/connect.tftest.hcl:154]: Fixed. This was a genuine module bug -- outputs.tf really did have insider_cidr_blocks instead of inside_cidr_blocks in the configurations output. Fixed the typo in 669c6294, and strengthened the test in 6e3acf36 to assert configurations[...].bgp_asn, .id, .peer_address, and .inside_cidr_blocks by name/value so a regression like this would fail loudly (an "Unsupported attribute" plan error, not just a silent pass) instead of just checking length().

File: modules/aws/transit_gateway/tgw/tests/tgw.tftest.hcl [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+  }
+
+  assert {
+    condition     = aws_ec2_transit_gateway.transit_gateway.tags["Name"] == "core-tgw"
+    error_message = "Name tag should default to the name variable."
+  }
+
+  assert {
+    condition     = output.id != null
+    error_message = "id output should resolve."
```
[the remark this answers is not in the document]
Comment (zachreborn) [reply modules/aws/transit_gateway/tgw/tests/tgw.tftest.hcl]: Fixed in 6e3acf36: the baseline run now asserts id, arn, bgp_asn, association_default_route_table_id, propagation_default_route_table_id, and transit_gateway_cidr_blocks all directly against the transit gateway resource's attributes.

File: modules/aws/transit_gateway/attachment/tests/attachment.tftest.hcl [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+  }
+
+  assert {
+    condition     = length(output.ids) == 2
+    error_message = "ids output should contain one entry per VPC."
+  }
+
+  assert {
+    condition     = length(output.info) == 2
+    error_message = "info output should contain one entry per VPC."
```
[the remark this answers is not in the document]
Comment (zachreborn) [reply modules/aws/transit_gateway/attachment/tests/attachment.tftest.hcl:161]: Fixed in 6e3acf36: now asserts info[...].id, .transit_gateway_id, .subnet_ids, and .appliance_mode_support against the attachment resource's actual attributes, and vpc_owner_id[...] against the resource's vpc_owner_id keyed by its real vpc_id, instead of only checking map length/non-nullness.

File: modules/aws/transit_gateway/attachment/tests/attachment.tftest.hcl [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+    error_message = "transit_gateway_default_route_table_association override should be honored."
+  }
+
+  assert {
+    condition     = aws_ec2_transit_gateway_vpc_attachment.this["transit_vpc"].transit_gateway_default_route_table_propagation == false
+    error_message = "transit_gateway_default_route_table_propagation override should be honored."
+  }
+}
+
+run "for_each_expands_to_one_attachment_per_vpc" {
```
[the remark this answers is not in the document]
Comment (zachreborn) [reply modules/aws/transit_gateway/attachment/tests/attachment.tftest.hcl:131]: Fixed in 6e3acf36: added run "empty_vpc_ids_map_creates_no_attachments" asserting zero attachments and empty ids/ids_list/info/vpc_owner_id outputs when vpc_ids = {}.

File: modules/aws/transit_gateway/connect/tests/connect.tftest.hcl [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+  }
+
+  assert {
+    condition     = aws_ec2_transit_gateway_connect_peer.peer["sdwan_vedge_1"].bgp_asn == "64512"
+    error_message = "bgp_asn should default to 64512."
+  }
+
+  assert {
+    condition     = aws_ec2_transit_gateway_connect_peer.peer["sdwan_vedge_1"].transit_gateway_address != null
+    error_message = "transit_gateway_address is Optional+Computed on the underlying resource -- when not set explicitly it should be left for AWS/the provider to assign, not coerced to a fixed value."
```
[the remark this answers is not in the document]
Comment (zachreborn) [reply modules/aws/transit_gateway/connect/tests/connect.tftest.hcl]: Fixed in 6e3acf36: added an override_resource block giving the peer a sentinel transit_gateway_address, and changed the assertion to check equality with that exact sentinel (instead of != null). override_resource only takes effect for attributes still null/computed in that run's plan, so if the module ever hard-coded a value for transit_gateway_address instead of passing each.value.transit_gateway_address through, the attribute would no longer be eligible for the override and this assertion would fail.
